File: platforms/ios/DezelScanner/DezelScanner/View/ObjectScannerView.swift
import AVFoundation
import Dezel

/**
 * @class ObjectScannerView
 * @since 0.1.0
 */
open class ObjectScannerView : UIView, CameraDelegate, ScannerDelegate {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * The content view delegate.
	 * @property delegate
	 * @since 0.1.0
	 */
	open weak var delegate: ObjectScannerViewDelegate?

	/**
	 * @property camera
	 * @since 0.1.0
	 * @hidden
	 */
	private var camera: Camera = Camera()

	/**
	 * @property detector
	 * @since 0.1.0
	 * @hidden
	 */
	private var scanner: Scanner = Scanner()

	/**
	 * @property capturing
	 * @since 0.1.0
	 * @hidden
	 */
	private var capturing: Bool = false

	/**
	 * @property scanning
	 * @since 0.1.0
	 * @hidden
	 */
	private var scanning: Bool = true

	/**
	 * @property started
	 * @since 0.1.0
	 * @hidden
	 */
	private var started: Bool = false

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	public override init(frame: CGRect) {

		super.init(frame: frame)

		self.camera.delegate = self
		self.scanner.delegate = self

		self.layer.addSublayer(self.camera.preview)

		self.addSubview(self.scanner.preview)
	}

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	required public init?(coder: NSCoder) {

		super.init(coder: coder)

		self.camera.delegate = self
		self.scanner.delegate = self

		self.layer.addSublayer(self.camera.preview)

		self.addSubview(self.scanner.preview)
	}

	/**
	 * @inherited
	 * @method layoutSubviews
	 * @since 0.1.0
	 */
	open override func layoutSubviews() {
		self.camera.preview.frame = self.bounds
		self.scanner.preview.frame = self.bounds
	}

	/**
	 * @method startCamera
	 * @since 0.1.0
	 */
	open func startCamera() {
		self.camera.start()
	}

	/**
	 * @method stopCamera
	 * @since 0.1.0
	 */
	open func stopCamera() {
		self.camera.stop()
	}

	/**
	 * @method startScanner
	 * @since 0.1.0
	 */
	open func startScanner() {
		if (self.scanning == false) {
			self.scanning = true
			self.scanner.enabled = true
		}
	}

	/**
	 * @method stopScanner
	 * @since 0.1.0
	 */
	open func stopScanner() {
		if (self.scanning) {
			self.scanning = false
			self.scanner.enabled = false
			self.delegate?.didLoseDocument(view: self)
			self.delegate?.didMissDocument(view: self)
		}
	}
	
	/**
	 * @method restartScanner
	 * @since 0.1.0
	 */
	open func restartScanner() {
		self.scanner.restart()
	}

	/**
	 * @method toggleFlash
	 * @since 0.1.0
	 */
	open func toggleFlash() {
		self.camera.toggleFlash()
	}

	/**
	 * @method captureImage
	 * @since 0.1.0
	 */
	open func captureImage() {
		self.capturing = true
	}

	//--------------------------------------------------------------------------
	// MARK: Methods - Animations
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method action
	 * @since 0.1.0
	 */
	override open func action(for layer: CALayer, forKey event: String) -> CAAction? {
		return Transition.action(for: layer, key: event)
	}

	//--------------------------------------------------------------------------
	// MARK: Camera Delegate
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method didCaptureFrame
	 * @since 0.1.0
	 */
	public func didCaptureFrame(camera: Camera, buffer: CMSampleBuffer) {

		if (self.started == false) {
			self.started = true
			DispatchQueue.main.async {
				self.delegate?.didActivate(view: self)
			}
		}

		if (self.capturing) {
			self.capturing = false
			if let image = CGImageCreateWithCMSampleBuffer(buffer) {
				DispatchQueue.main.async {
					self.delegate?.didCaptureImage(view: self, image: UIImage(cgImage: image))
				}
			}
		}

		self.scanner.process(buffer)
	}

	/**
	 * @inherited
	 * @method didStartMoving
	 * @since 0.1.0
	 */
	public func didStartMoving(camera: Camera) {
		if (self.scanning) {
			self.scanner.enabled = false
			self.scanner.reset()
		}
	}

	/**
	 * @inherited
	 * @method didStopMoving
	 * @since 0.1.0
	 */
	public func didStopMoving(camera: Camera) {
		if (self.scanning) {
			self.scanner.enabled = true
			self.scanner.reset()
		}
	}

	//--------------------------------------------------------------------------
	// MARK: Scanner Delegate
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method didFindObject
	 * @since 0.1.0
	 */
	public func didFindObject(detector: Scanner, shape: [CGPoint], image: UIImage) {
		self.delegate?.didFindDocument(view: self, shape: self.transform(shape), image: image)
	}

	/**
	 * @inherited
	 * @method didSpotObject
	 * @since 0.1.0
	 */
	public func didSpotObject(detector: Scanner, shape: [CGPoint]) {
		self.delegate?.didSpotDocument(view: self, shape: self.transform(shape))
	}

	/**
	 * @inherited
	 * @method didLoseObject
	 * @since 0.1.0
	 */
	public func didLoseObject(detector: Scanner) {
		self.delegate?.didLoseDocument(view: self)
	}

	/**
	 * @inherited
	 * @method didMissObject
	 * @since 0.1.0
	 */
	public func didMissObject(detector: Scanner) {
		self.delegate?.didMissDocument(view: self)
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
	 * @method transform
	 * @since 0.1.0
	 * @hidden
	 */
	private func transform(_ points: [CGPoint]) -> [CGPoint] {

		let bw = self.scanner.sampleBufferWidth
		let bh = self.scanner.sampleBufferHeight

		let cx = bw / 2
		let cy = bh / 2

		var t = CGAffineTransform.identity

		if (self.scanner.debug) {
			t = t.translatedBy(x: +cx, y: +cy)
			t = t.scaledBy(x: bh / self.camera.preview.bounds.height, y: 1)
			t = t.translatedBy(x: -cx, y: -cy)
		}

		var result = [CGPoint]()
		result.append(self.camera.preview.layerPointConverted(fromCaptureDevicePoint: CGPoint(x: points[0].y / bh, y: 1 - points[0].x / bw)).applying(t))
		result.append(self.camera.preview.layerPointConverted(fromCaptureDevicePoint: CGPoint(x: points[1].y / bh, y: 1 - points[1].x / bw)).applying(t))
		result.append(self.camera.preview.layerPointConverted(fromCaptureDevicePoint: CGPoint(x: points[2].y / bh, y: 1 - points[2].x / bw)).applying(t))
		result.append(self.camera.preview.layerPointConverted(fromCaptureDevicePoint: CGPoint(x: points[3].y / bh, y: 1 - points[3].x / bw)).applying(t))

		return result
	}
}

File: platforms/android/DezelScanner/lib/src/main/java/ca/logaritm/dezel/scanner/view/ObjectScannerView.kt
package ca.logaritm.dezel.scanner.view

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.graphics.PointF
import android.media.Image
import android.util.Log
import android.view.ViewGroup
import ca.logaritm.dezel.scanner.camera.Camera
import ca.logaritm.dezel.scanner.camera.CameraListener
import ca.logaritm.dezel.scanner.camera.view.CameraView
import ca.logaritm.dezel.scanner.scanner.Scanner
import ca.logaritm.dezel.scanner.scanner.ScannerListener
import ca.logaritm.dezel.view.graphic.Convert

open class ObjectScannerView(context: Context) : ViewGroup(context), CameraListener, ScannerListener {

	//--------------------------------------------------------------------------
	// Static
	//--------------------------------------------------------------------------

	companion object {
		init {
			System.loadLibrary("dezel_scanner")
		}
	}

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * The object scanner listener.
	 * @property listener
	 * @since 0.1.0
	 */
	open var listener: ObjectScannerViewListener? = null

	/**
	 * @property camera
	 * @since 0.1.0
	 * @hidden
	 */
	private var camera: Camera = Camera(context)

	/**
	 * @property scanner
	 * @since 0.1.0
	 * @hidden
	 */
	private var scanner: Scanner = Scanner(context)

	/**
	 * @property capturing
	 * @since 0.1.0
	 * @hidden
	 */
	private var capturing: Boolean = false

	/**
	 * @property scanning
	 * @since 0.1.0
	 * @hidden
	 */
	private var scanning: Boolean = true

	/**
	 * @property started
	 * @since 0.1.0
	 * @hidden
	 */
	private var started: Boolean = false

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	init {

		this.addView(this.camera.preview)
		this.addView(this.scanner.preview)
		this.camera.listener = this
		this.scanner.listener = this
		//this.scanner.preview.alpha = 0.75f

		this.scanner.debug = true
	}

	/**
	 * @method startCamera
	 * @since 0.1.0
	 * @hidden
	 */
	open fun startCamera() {
		this.camera.start()
	}

	/**
	 * @method stopCamera
	 * @since 0.1.0
	 * @hidden
	 */
	open fun stopCamera() {
		this.camera.stop()
	}

	/**
	 * @method startScanner
	 * @since 0.1.0
	 */
	open fun startScanner() {
		if (this.scanning == false) {
			this.scanning = true
			this.scanner.enabled = true
		}
	}

	/**
	 * @method stopScanner
	 * @since 0.1.0
	 */
	open fun stopScanner() {
		if (this.scanning) {
			this.scanning = false
			this.scanner.enabled = false
			this.listener?.onLoseDocument(this)
			this.listener?.onMissDocument(this)
		}
	}

	/**
	 * @method restartScanner
	 * @since 0.1.0
	 */
	open fun restartScanner() {
		this.scanner.restart()
	}

	/**
	 * @method toggleFlash
	 * @since 0.1.0
	 */
	open fun toggleFlash() {
		this.camera.toggleFlash()
	}

	/**
	 * @method captureImage
	 * @since 0.1.0
	 */
	open fun captureImage() {
		this.capturing = true
	}

	/**
	 * @inherited
	 * @method onMeasure
	 * @since 0.1.0
	 */
	override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
		super.onMeasure(widthMeasureSpec, heightMeasureSpec)
		this.camera.preview.measure(widthMeasureSpec, heightMeasureSpec)
		this.scanner.preview.measure(widthMeasureSpec, heightMeasureSpec)
	}

	/**
	 * @inherited
	 * @method onLayout
	 * @since 0.1.0
	 */
	override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
		this.camera.preview.layout(l, t, r, b)
		this.scanner.preview.layout(l, t, r, b)
	}

	//--------------------------------------------------------------------------
	// Camera Listener
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method onCaptureFrame
	 * @since 0.1.0
	 */
	override fun onCaptureFrame(camera: Camera, frame: Image) {

		if (this.started == false) {
			this.started = true
			this.post {
				this.listener?.onActivate(this)
			}
		}

		if (this.capturing) {
			this.capturing = false

			val bitmap = CameraView.convert(frame)
			if (bitmap != null) {

				val matrix = Matrix()
				matrix.postRotate(90f)
				val rotated = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)

				this.post {
					this.listener?.onCaptureImage(this, rotated)
				}
			}

			return
		}

		this.scanner.process(frame)
	}

	/**
	 * @inherited
	 * @method onStartMoving
	 * @since 0.1.0
	 */
	override fun onStartMoving(camera: Camera) {
		if (this.scanning) {
			this.scanner.enabled = false
			this.scanner.reset()
		}
	}

	/**
	 * @inherited
	 * @method onStopMoving
	 * @since 0.1.0
	 */
	override fun onStopMoving(camera: Camera) {
		if (this.scanning) {
			this.scanner.enabled = true
			this.scanner.reset()
		}
	}

	//--------------------------------------------------------------------------
	// Scanner Listener
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method onFindObject
	 * @since 0.1.0
	 */
	override fun onFindObject(detector: Scanner, shape: Array<PointF>, image: Bitmap) {
		this.listener?.onFindDocument(this, this.transform(shape), image)
	}

	/**
	 * @inherited
	 * @method onSpotObject
	 * @since 0.1.0
	 */
	override fun onSpotObject(detector: Scanner, shape: Array<PointF>) {
		this.listener?.onSpotDocument(this, this.transform(shape))
	}

	/**
	 * @inherited
	 * @method onLoseObject
	 * @since 0.1.0
	 */
	override fun onLoseObject(detector: Scanner) {
		this.listener?.onLoseDocument(this)
	}

	/**
	 * @inherited
	 * @method onMissObject
	 * @since 0.1.0
	 */
	override fun onMissObject(detector: Scanner) {
		this.listener?.onMissDocument(this)
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	/**
	 * @method transform
	 * @since 0.1.0
	 * @hidden
	 */
	private fun transform(points: Array<PointF>): Array<PointF> {

		val transformed = mutableListOf<PointF>()

		for (point in points) {
			val p = this.camera.convert(point)
			transformed.add(PointF(
				Convert.toDp(p.x),
				Convert.toDp(p.y)
			))
		}

		return transformed.toTypedArray()
	}
}
